### PR TITLE
make: unify test target

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

Most tests had the same two lines for ```test:``` identical.
This PR moves those lines to a central location, adding some flexibility options for the corner cases.

### CODE REVIEW

**Comment by kaspar030** [comment]
@miri64 can you remember the reason for ```TERMFLAGS=``` for the trace test? It works without, is it still necessary?

**Comment by kaspar030** [comment]
(force-pushed an accidental semantic change in tests/od)

File: Makefile.include [hunk 1]
```diff
@@ -478,6 +478,10 @@ reset:
 	$(call check_cmd,$(RESET),Reset program)
 	$(RESET) $(RESET_FLAGS)
 
+TESTS ?= $(wildcard $(APPDIR)/tests/*)
```
Comment (smlng) [inline Makefile.include]: can we make this a bit more specific, i.e., match on python scripts - just in case there are other files or subdirs that are not supposed to be run

**Comment by miri64** [comment]
I remember to have something similar prepared, but I didn't pull through for a specific reason. I don't remember why exactly, but maybe I find the reason again over the weekend...

> @miri64 can you remember the reason for TERMFLAGS= for the trace test? It works without, is it still necessary?

I think, and `git log -S TERMFLAGS -p -- tests/*/Makefile` seem to confirm that, this is just an artifact from when #8448 wasn't merged yet that slipped through the review of that test.

**Comment by kaspar030** [comment]
> I think, and git log -S TERMFLAGS -p -- tests/*/Makefile seem to confirm that, this is just an artifact from when #8448 wasn't merged yet that slipped through the review of that test.

Very good, my first tries of passing variables to the tests looked quite ugly. ;)

**Comment by cladmi** [comment]
It removes one thing that before having a `test` target meant there are indeed tests to run and so it is required to flash the board and try running them. I used it to know if I should run test or not.
In the current state, this is not available anymore and checking if `TESTS` is not empty is also not enough as the check for executable files is done in `test` and not in the variable definition.

However, having the `TESTS` variable is nice because it goes in the direction of having them listed and so available to be used as dependencies for maybe not re-run tests.

**Comment by cladmi** [comment]
I get the behavior I prefer by moving `test -x` in the variable definition.
Like for example with the following diff:

``` diff
diff --git a/Makefile.include b/Makefile.include
index b882a9a..95b844a 100644
--- a/Makefile.include
+++ b/Makefile.include
@@ -478,10 +478,10 @@ reset:
        $(call check_cmd,$(RESET),Reset program)
        $(RESET) $(RESET_FLAGS)
 
-TESTS ?= $(wildcard $(APPDIR)/tests/*)
+TESTS ?= $(foreach file,$(wildcard $(APPDIR)/tests/*),\
+                        $(shell test -f $(file) -a -x $(file) && echo $(file)))
 test: $(TEST_DEPS)
        for t in $(TESTS); do \
-               test -f $$t -a -x $$t || continue ; \
                $$t || exit 1; \
        done
 
```

I tested with some crap files:

```
ls tests/unittests/tests
01-run.py  directory  README

make --no-print-directory -C tests/unittests/ info-debug-variable-TESTS
/home/harter/work/git/RIOT/tests/unittests/tests/01-run.py
```

I would ACK with a change doing this.

**Comment by kaspar030** [comment]
> I get the behavior I prefer by moving test -x in the variable definition.

Does that make the subshells running ```test -x``` *always* execute, compared to only when execeuting the test: recipe?

**Comment by cladmi** [comment]
> Does that make the subshells running test -x always execute, compared to only when execeuting the test: recipe?

No as the `TESTS` is defined with `?=` it is a deferred variable and only executed at evaluation.

**Comment by gebart** [comment]
Doesn't make have anything similar to test -x without sub shell?

**Comment by kaspar030** [comment]
> Doesn't make have anything similar to test -x without sub shell?

Can't find anything. ;(

- rebased, addressed @cladmi's comment

File: Makefile.include [hunk 2]
```diff
@@ -536,10 +536,10 @@ reset:
 	$(call check_cmd,$(RESET),Reset program)
 	$(RESET) $(RESET_FLAGS)
 
-TESTS ?= $(wildcard $(APPDIR)/tests/*)
+TESTS ?= $(foreach file,$(wildcard $(APPDIR)/tests/*),\
+                        $(shell test -f $(file) -a -x $(file) && echo $(file)))
 test: $(TEST_DEPS)
 	for t in $(TESTS); do \
```
Comment (cladmi) [inline Makefile.include]: Maybe silent this also with `$(Q)`

**Comment by cladmi** [comment]
Still a small issue,`tests/od` only finds one test, I try to find  a solution for it.

Otherwise everything looks good.

Tested:

```
$ make --silent --no-print-directory -C examples/hello-world/ info-debug-variable-TESTS

$ make --silent --no-print-directory -C tests/unittests/ info-debug-variable-TESTS
/home/cladmi/git/work/RIOT/tests/unittests/tests/01-run.py

```

`make test` does nothing in `examples/hello-world` and runs a test in `tests/bloom_bytes`.

**Comment by cladmi** [comment]
Oh, my bad >< it is overwritten in the `tests/od/Makefile`. So everything is fine.

**Comment by cladmi** [comment]
Ready to squash for me.

**Comment by kaspar030** [comment]
> Ready to squash for me.

did.

**Comment by kaspar030** [comment]
@smlng you fine with the solution to your comment?

File: tests/pkg_fatfs_vfs/Makefile [hunk 3]
```diff
@@ -43,8 +43,10 @@ BOARD_WHITELIST := airfy-beacon arduino-due arduino-duemilanove arduino-mega2560
                    nz32-sc151 openmote-cc2538 pba-d-01-kw2x remote-pa remote-reva \
                    remote-revb samd21-xpro saml21-xpro samr21-xpro sltb001a \
                    sodaq-autonomo sodaq-explorer spark-core stm32f0discovery \
-                   stm32f3discovery stm32f4discovery udoo waspmote-pro \
-                   yunjia-nrf51822 native
+                   stm32f3discovery stm32f4discovery telosb udoo waspmote-pro \
```
Comment (cladmi) [inline tests/pkg_fatfs_vfs/Makefile]: These BOARD_WHITELIST changes look unrelated, please remove them, you can directly squash as they are easy to review.

Comment (kaspar030) [reply tests/pkg_fatfs_vfs/Makefile]: Yup, those were a rebase mistake.

**Comment by cladmi** [comment]
There are also some new tests added since last time

```
git grep '^test:' -- examples/ tests/
tests/libc_newlib/Makefile:test:
tests/pkg_monocypher/Makefile:test:
```

```
make -C tests/libc_newlib/ BOARD=samr21-xpro all flash test
make: Entering directory '/home/cladmi/git/work/RIOT/tests/libc_newlib'
Makefile:47: warning: overriding recipe for target 'test'
/home/cladmi/git/work/RIOT/tests/libc_newlib/../../Makefile.include:542: warning: ignoring old recipe for target 'test'
```

**Comment by kaspar030** [comment]
> There are also some new tests added since last time

fixed

**Comment by cladmi** [comment]
Re-ACK Good for me, you can squash this one.

This will need to be done for some time to check newly added 'test' commands in already existing PRs.

@smlng something to say before I merge it ?

**Comment by kaspar030** [comment]
and squashed. @smlng IMO your comment has been addressed?
